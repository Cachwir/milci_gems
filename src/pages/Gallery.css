.gallery .container.floated  .gallery-images-selector-container {
    padding: 0;
    width: 300px;
}
.gallery .container.floated .gallery-current-element {
    padding: 0;
    width: 720px;
}
.gallery .top-image .post-img {
    width: 100%;
    height: 350px;
    /*padding-top: 44.11%; !* 34:15 Aspect Ratio *!*/
    /*position: relative; !* If you want text inside of it *!*/
    /*overflow: hidden;*/
}
.gallery .top-image .post-img img {
    max-height: 100%;
    margin: auto;
}
.gallery .gallery-current-element .post {
    margin: 40px;
}
.gallery .gallery-images-selector {
    max-height: 600px;
    overflow-y : auto;
    padding: 20px 30px;
}

@media only screen and (max-width: 767px){
    .gallery .gallery-current-element .post {
        margin: 24px;
    }
    .gallery .container.floated  .gallery-images-selector-container {
        width: 100%;
    }
}
