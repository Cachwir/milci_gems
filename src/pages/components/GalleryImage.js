import React from 'react';
import PropTypes from 'prop-types';

const GalleryImage = ({ path, title, subTitle, text, source }) => (
    <div>
        <article className="top-image">
            <figure className="post-img picture">
                    <img src={path} alt={title} />
            </figure>
        </article>
        <article className="post">
            <h3 className="margin">{title}</h3>
            <span>{subTitle}</span>
            <p>{text}</p>
            <div className="right">
                <a href={source} target="_blank" rel="noopener noreferrer">source</a>
            </div>
        </article>
    </div>
);

GalleryImage.propTypes = {
    path: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    subTitle: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
    source: PropTypes.string.isRequired,
};

export default GalleryImage;