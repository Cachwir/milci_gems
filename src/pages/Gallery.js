import React, {Component} from "react";
import {Route} from "react-router-dom";
import { withRouter } from "react-router";
import {AnimatedSwitch} from "react-router-transition";
import axios from 'axios';
import GalleryImage from "./components/GalleryImage";
import GalleryImagesSelector from "./components/GalleryImagesSelector";

import "./Gallery.css";
import {localStorageGet, localStorageSet} from "../lib/functions";
import Config from "../lib/Config";

class GalleryWithoutRouter extends Component
{
    FETCH_POST_ROUTE    = "posts/";
    PATH_TO_POST_IMAGES = "/assets/dynamic/postImages/";

    constructor() {
        super();
        this.localStorageCurrentImageIdKey = "currentImageId";
        this.defaultImageId = 0;
        this.currentImageId = this.getCurrentImageId();
        this.state = {
            images: [],
        }
    }

    getFetchPostUrl() {
        return Config.getBackendUrl() + this.FETCH_POST_ROUTE;
    }

    getImageUrl = (imageId) =>
    {
        const { url } = this.props.match;
        return url + "/" + imageId + "-" + encodeURI(this.state.images[imageId].title
            .replace(" ", "")
            .replace("é", "e")
            .replace("è", "e")
            .replace("à", "a")
            .replace("ù", "u")
        );
    };

    componentDidMount()
    {
        // Retrieves the posts
        this.fetchPosts()
            .then(response =>
            {
                if (typeof response.data !== "undefined" && response.data.length > 0)
                {
                    this.addImages(response.data);
                }
            })
            .catch((error) =>
            {
                console.log(error);
            })
    }

    fetchPosts()
    {
        // Retrieves the posts
        return axios.get(this.getFetchPostUrl());
    }

    // fx binding to conserve this
    addImages = (imagesData) =>
    {
        const images = [];

        imagesData.forEach((imageData) =>
        {
            images.push({
                title: imageData.title,
                subTitle: imageData.subtitle,
                path: this.PATH_TO_POST_IMAGES + imageData.image,
                text: imageData.text,
                source: imageData.source,
                getAccessUrl: this.getImageUrl,
            });
        });

        this.setState({
            images: images
        });
    };

    getCurrentImageId()
    {
        return localStorageGet(this.localStorageCurrentImageIdKey, this.defaultImageId);
    }

    setCurrentImageId(newImageId)
    {
        this.currentImageId = newImageId;
        localStorageSet(this.localStorageCurrentImageIdKey, newImageId);
    }

    // keeping this on fx binding
    onRouteChanged = (newCurrentImageId) =>
    {
        if (this.currentImageId !== newCurrentImageId)
        {
            this.setCurrentImageId(newCurrentImageId);
        }
    };

    render()
    {
        const doesFirstImageExist = typeof this.state.images[this.currentImageId] !== "undefined";

        return (
            <section className="gallery">
                <div id="content">

                    <div className="container floated">

                        {/* Page Content */}
                        <div className="gallery-current-element eleven floated left">
                            <AnimatedSwitch
                                atEnter={{
                                    opacity: 0,
                                    marginLeft: -100,
                                }}
                                atLeave={{
                                    opacity: 0,
                                    marginLeft: 0,
                                }}
                                atActive={{
                                    opacity: 1,
                                    marginLeft: 0,
                                }}
                                className="switch-wrapper"
                            >
                                {/* A <Switch> looks through its children <Route>s and
                                 renders the first one that matches the current URL. */}
                                {doesFirstImageExist && (
                                    <Route key={-1} exact path={"/gallery"}>
                                        <GalleryImage {...this.state.images[this.currentImageId]} />
                                    </Route>
                                )}
                                {this.state.images.map((galleryImage, index) => (
                                    <Route key={index} exact path={this.getImageUrl(index)}>
                                        <GalleryImage {...galleryImage} />
                                    </Route>
                                ))}
                            </AnimatedSwitch>
                        </div>
                        {/* Page Content / End */}

                        {/* Sidebar */}
                        <div className="gallery-images-selector-container four floated sidebar right">
                            <GalleryImagesSelector images={this.state.images} onThumbnailClick={this.onRouteChanged}/>
                        </div>
                        {/* Sidebar / End */}
                    </div>

                </div>
            </section>
        );
    }
}

// Create a new component that is "connected" (to borrow redux
// terminology) to the router.
const Gallery = withRouter(GalleryWithoutRouter);

export default Gallery;