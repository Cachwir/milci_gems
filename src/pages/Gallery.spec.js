import { shallow } from 'enzyme'
import { expect } from 'chai'
import React from 'react'
import {Route} from "react-router-dom";
import axios from 'axios';
import Gallery from "./Gallery";
import GalleryImage from "./components/GalleryImage";

const prepareMemoryRouter = () =>
{
    axios.get.mockImplementationOnce(() => Promise.resolve({}));

    const wrapper = shallow(
        <Gallery.WrappedComponent />
    );
    wrapper.setProps({
        match: "https://my_fake_url.com",
    });

    wrapper.instance().addImages(generateFakeImages());

    return wrapper;
};

const generateFakeImages = (quantity = 5) =>
{
    const images = [];

    for (let i = 0; i < quantity; i++)
    {
        images.push({
            title: "title-" + i,
            subtitle: "subtitle-" + i,
            text: "text-" + i,
            image: "image-" + i + ".jpg",
            source: "#",
        });
    }

    return images;
};

jest.mock('axios');

describe('<Gallery />', () =>
{
    it('renders without crashing', () => {
        const wrapper = prepareMemoryRouter();
        expect(wrapper.find(GalleryImage).exists()).to.equal(true);
    });

    it('has 6 routes', () => {
        const wrapper = prepareMemoryRouter();
        expect(wrapper.find(Route)).to.has.length(6);
    });

    it('should match its reference snapshot', () => {
        const wrapper = prepareMemoryRouter();
        expect(wrapper).to.matchSnapshot();
    });
});