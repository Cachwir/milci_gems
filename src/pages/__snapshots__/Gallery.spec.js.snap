// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Gallery /> should match its reference snapshot 1`] = `
<section
  className="gallery"
>
  <div
    id="content"
  >
    <div
      className="container floated"
    >
      <div
        className="gallery-current-element eleven floated left"
      >
        <AnimatedSwitch
          atActive={
            Object {
              "marginLeft": 0,
              "opacity": 1,
            }
          }
          atEnter={
            Object {
              "marginLeft": -100,
              "opacity": 0,
            }
          }
          atLeave={
            Object {
              "marginLeft": 0,
              "opacity": 0,
            }
          }
          className="switch-wrapper"
        >
          <Route
            exact={true}
            key="-1"
            path="/gallery"
          >
            <GalleryImage
              getAccessUrl={[Function]}
              path="/assets/dynamic/postImages/image-0.jpg"
              source="#"
              subTitle="subtitle-0"
              text="text-0"
              title="title-0"
            />
          </Route>
          <Route
            exact={true}
            key="0"
            path="undefined/0-title-0"
          >
            <GalleryImage
              getAccessUrl={[Function]}
              path="/assets/dynamic/postImages/image-0.jpg"
              source="#"
              subTitle="subtitle-0"
              text="text-0"
              title="title-0"
            />
          </Route>
          <Route
            exact={true}
            key="1"
            path="undefined/1-title-1"
          >
            <GalleryImage
              getAccessUrl={[Function]}
              path="/assets/dynamic/postImages/image-1.jpg"
              source="#"
              subTitle="subtitle-1"
              text="text-1"
              title="title-1"
            />
          </Route>
          <Route
            exact={true}
            key="2"
            path="undefined/2-title-2"
          >
            <GalleryImage
              getAccessUrl={[Function]}
              path="/assets/dynamic/postImages/image-2.jpg"
              source="#"
              subTitle="subtitle-2"
              text="text-2"
              title="title-2"
            />
          </Route>
          <Route
            exact={true}
            key="3"
            path="undefined/3-title-3"
          >
            <GalleryImage
              getAccessUrl={[Function]}
              path="/assets/dynamic/postImages/image-3.jpg"
              source="#"
              subTitle="subtitle-3"
              text="text-3"
              title="title-3"
            />
          </Route>
          <Route
            exact={true}
            key="4"
            path="undefined/4-title-4"
          >
            <GalleryImage
              getAccessUrl={[Function]}
              path="/assets/dynamic/postImages/image-4.jpg"
              source="#"
              subTitle="subtitle-4"
              text="text-4"
              title="title-4"
            />
          </Route>
        </AnimatedSwitch>
      </div>
      <div
        className="gallery-images-selector-container four floated sidebar right"
      >
        <GalleryImagesSelector
          images={
            Array [
              Object {
                "getAccessUrl": [Function],
                "path": "/assets/dynamic/postImages/image-0.jpg",
                "source": "#",
                "subTitle": "subtitle-0",
                "text": "text-0",
                "title": "title-0",
              },
              Object {
                "getAccessUrl": [Function],
                "path": "/assets/dynamic/postImages/image-1.jpg",
                "source": "#",
                "subTitle": "subtitle-1",
                "text": "text-1",
                "title": "title-1",
              },
              Object {
                "getAccessUrl": [Function],
                "path": "/assets/dynamic/postImages/image-2.jpg",
                "source": "#",
                "subTitle": "subtitle-2",
                "text": "text-2",
                "title": "title-2",
              },
              Object {
                "getAccessUrl": [Function],
                "path": "/assets/dynamic/postImages/image-3.jpg",
                "source": "#",
                "subTitle": "subtitle-3",
                "text": "text-3",
                "title": "title-3",
              },
              Object {
                "getAccessUrl": [Function],
                "path": "/assets/dynamic/postImages/image-4.jpg",
                "source": "#",
                "subTitle": "subtitle-4",
                "text": "text-4",
                "title": "title-4",
              },
            ]
          }
          onThumbnailClick={[Function]}
        />
      </div>
    </div>
  </div>
</section>
`;
